src/core/thread/mle_router.cpp

```diff
@@ -3484,20 +3484,66 @@ uint16_t MleRouter::GetNextHop(uint16_t aDestination)
     return rval;
 }
 
-uint8_t MleRouter::GetCost(uint16_t aRloc16)
+uint8_t MleRouter::GetPathCost(uint16_t aDestRloc16) const
 {
-    uint8_t routerId = RouterIdFromRloc16(aRloc16);
-    uint8_t cost     = GetLinkCost(routerId);
-    Router *router   = mRouterTable.FindRouterById(routerId);
-    uint8_t routeCost;
+    uint8_t       cost = kMaxRouteCost;
+    uint8_t       destRouterId;
+    const Router *router;
+    const Router *nextHop;
 
-    VerifyOrExit(router != nullptr && mRouterTable.FindNextHopOf(*router) != nullptr);
+    if (aDestRloc16 == GetRloc16())
+    {
+        // Destination is this device, return cost as zero.
+        // This is also valid when device is a child.
+        ExitNow(cost = 0);
+    }
+
+    if (IsChild())
+    {
+        // If device is a child, then check if destination is our parent
+        // and determine cost based on the link quality to parent.
+        // Otherwise we cannot determine the cost.
+
+        VerifyOrExit(aDestRloc16 == GetParent().GetRloc16());
+        cost = CostForLinkQuality(GetParent().GetLinkQualityIn());
+        ExitNow();
```

Comment:
One idea (possibly for future enahancment) on an `FED`, if we can store our parent's `RouteTlv` (route info), we can use it to calculate the cost and make a better decision (if we have option of multiple router BRs). Another case is if the dest happens to be another child of our parent. I did not add it in the code now though (we cannot know  the cost/link-quality from parent to the child BR, we can estimate/assume it being `q3` or `q2`). @jwhui any ideas/suggestions?

Reply:
I think it makes sense for a FED to process Route TLVs and use that information. Certainly a REED should be able to process Route TLVs. Technically, a FED does not need to maintain a routing table, but I don't think we make that distinction in OpenThread.

Also, I agree we should also optimize for child attached to the same parent.

Reply:
Thanks. Will add both in a follow-up (next) PR.